### ISSUE TITLE
No digest available for image 

### ISSUE THREAD

**Issue by mastoj** [pulumi/pulumi-docker#245]
I have a weird error. Running my build an publish locally yields different results compared to on Azure DevOps. I'm using the exact same build scripts for doing it locally as on the DevOps. The difference is that it works on DevOps but fails locally with the error:

```
    error: Pulumi.ResourceException: No digest available for image <image name>:41c9f8
       at Pulumi.Docker.Docker.BuildImageAsync(String imageName, Union`2 pathOrBuild, Resource logResource, String[] cacheFrom)
       at Pulumi.Docker.Docker.BuildAndPushImageWorkerAsync(String baseImageName, Union`2 pathOrBuild, String repositoryUrl, Resource logResource, Boolean skipPush, ImageRegistryUnwrap registry)
       at Pulumi.Docker.Docker.<>c__DisplayClass0_0.<<BuildAndPushImageAsync>b__0>d.MoveNext()
    --- End of stack trace from previous location ---
       at Pulumi.Output`1.ApplyHelperAsync[U](Task`1 dataTask, Func`2 func)
       at Pulumi.Output`1.ApplyHelperAsync[U](Task`1 dataTask, Func`2 func)
       at Pulumi.Output`1.Pulumi.IOutput.GetDataAsync()
       at Pulumi.Serialization.Serializer.SerializeAsync(String ctx, Object prop, Boolean keepResources)
       at Pulumi.Deployment.SerializeFilteredPropertiesAsync(String label, IDictionary`2 args, Predicate`1 acceptKey, Boolean keepResources)
       at Pulumi.Deployment.SerializeAllPropertiesAsync(String label, IDictionary`2 args, Boolean keepResources)
       at Pulumi.Deployment.RegisterResourceOutputsAsync(Resource resource, Output`1 outputs)
       at Pulumi.Deployment.Runner.<>c__DisplayClass7_0.<<WhileRunningAsync>g__HandleCompletion|0>d.MoveNext()
    --- End of stack trace from previous location ---
       at Pulumi.Deployment.Runner.WhileRunningAsync()

``` 

Not sure what the "No digest available for image" mean. The way I connect to the container registry is asking another stack for the credentials, and that is exactly the same code running locally as on DevOps.

**Comment by mastoj** [comment]
Reading the code here, https://github.com/pulumi/pulumi-docker/blob/master/sdk/dotnet/Docker.cs/#L365-L374, it seems like the image after it has been built isn't available.

**Comment by leezen** [comment]
This sounds like https://github.com/pulumi/pulumi-docker/issues/200 which was previously fixed. Are you on the latest version of the package?

**Comment by chwunder** [comment]
hi everyone, unfortunately I also do receive the same error.
Diagnostics:
  pulumi:pulumi:Stack (test-deploy-isolatedenvintegration):
    error: Program failed with an unhandled exception:
    error: Traceback (most recent call last):
      File "/pulumi/bin/pulumi-language-python-exec", line 107, in <module>
        loop.run_until_complete(coro)
      File "/usr/local/lib/python3.9/asyncio/base_events.py", line 647, in run_until_complete
        return future.result()
      File "/usr/local/lib/python3.9/site-packages/pulumi/runtime/stack.py", line 126, in run_in_stack
        await run_pulumi_func(lambda: Stack(func))
      File "/usr/local/lib/python3.9/site-packages/pulumi/runtime/stack.py", line 51, in run_pulumi_func
        await wait_for_rpcs()
      File "/usr/local/lib/python3.9/site-packages/pulumi/runtime/stack.py", line 110, in wait_for_rpcs
        raise exception
      File "/usr/local/lib/python3.9/site-packages/pulumi/output.py", line 98, in is_value_known
        return await is_known and not contains_unknowns(await future)
      File "/usr/local/lib/python3.9/site-packages/pulumi/output.py", line 214, in run
        return await cast(Awaitable[U], transformed)
      File "/usr/local/lib/python3.9/site-packages/pulumi_docker/docker.py", line 243, in build_and_push_image
        build_result = await build_image(base_image_name, path_or_build, log_resource, cache_from)
      File "/usr/local/lib/python3.9/site-packages/pulumi_docker/docker.py", line 444, in build_image
        raise ResourceError(
    pulumi_docker.docker.ResourceError: No digest available for image registry7b95ac5f.azurecr.io/extollo-instance
    error: an unhandled error occurred: Program exited with non-zero exit code: 1

pulumi              3.29.1
pulumi-azure-native 1.62.0
pulumi-docker       3.2.0

**Comment by rodrigoalmeida94** [comment]
@leezen @chwunder I also encountered the same issue. Could it be that the fix in https://github.com/pulumi/pulumi-docker/issues/200 hasn't been applied for the python SDK? If I understand correctly, the PR addresses the issue in the .NET SDK.

These are my package versions:
```
pulumi 3.33.1
pulumi-gcp 6.24.0
pulumi-docker 3.2.0
```

**Comment by Gwojda** [comment]
Hi,
Never had the problem before, but now 'got this issue every time i run pulumi up.
Any news?

**Comment by DJRHails** [comment]
Can confirm I'm also experiencing this error in python:
```
pulumi==3.35.1
pulumi-docker==3.2.0
pulumi-gcp==6.28.0
```

**Comment by Gwojda** [comment]
Hi @DJRHails,
Quick fix is to set pulumi-docker==3.1
pulumi-docker 3.2 is broken.

**Comment by dixler** [comment]
Ran into this bug when working on https://github.com/pulumi/automation-api-examples/pull/39 if someone needs a repro.

**Comment by prudnikov** [comment]
pulumi-3.42.0 
pulumi-aws-5.16.2 
pulumi-docker-3.4.1
I have the same issue in TeamCity build, everything is fine when running `pulumi up` locally. 


```
pulumi:pulumi:Stack (XXXXXX-XXX-XXXXXXX):
[07:30:20]     error: Program failed with an unhandled exception:
[07:30:20]     Traceback (most recent call last):
[07:30:20]       File "/pulumi/bin/pulumi-language-python-exec", line 179, in <module>
[07:30:20]         loop.run_until_complete(coro)
[07:30:20]       File "/usr/local/lib/python3.9/asyncio/base_events.py", line 647, in run_until_complete
[07:30:20]         return future.result()
[07:30:20]       File "/teamcity/work/56a47d90f832aeb3/venv/lib/python3.9/site-packages/pulumi/runtime/stack.py", line 126, in run_in_stack
[07:30:20]         await run_pulumi_func(lambda: Stack(func))
[07:30:20]       File "/teamcity/work/56a47d90f832aeb3/venv/lib/python3.9/site-packages/pulumi/runtime/stack.py", line 51, in run_pulumi_func
[07:30:20]         await wait_for_rpcs()
[07:30:20]       File "/teamcity/work/56a47d90f832aeb3/venv/lib/python3.9/site-packages/pulumi/runtime/stack.py", line 73, in wait_for_rpcs
[07:30:20]         await RPC_MANAGER.rpcs.pop()
[07:30:20]       File "/teamcity/work/56a47d90f832aeb3/venv/lib/python3.9/site-packages/pulumi/runtime/rpc_manager.py", line 68, in rpc_wrapper
[07:30:20]         result = await rpc
[07:30:20]       File "/teamcity/work/56a47d90f832aeb3/venv/lib/python3.9/site-packages/pulumi/runtime/resource.py", line 685, in do_register_resource_outputs
[07:30:20]         serialized_props = await rpc.serialize_properties(outputs, {})
[07:30:20]       File "/teamcity/work/56a47d90f832aeb3/venv/lib/python3.9/site-packages/pulumi/runtime/rpc.py", line 207, in serialize_properties
[07:30:20]         result = await serialize_property(
[07:30:20]       File "/teamcity/work/56a47d90f832aeb3/venv/lib/python3.9/site-packages/pulumi/runtime/rpc.py", line 375, in serialize_property
[07:30:20]         is_known = await output._is_known
[07:30:20]       File "/teamcity/work/56a47d90f832aeb3/venv/lib/python3.9/site-packages/pulumi/output.py", line 100, in is_value_known
[07:30:20]         return await is_known and not contains_unknowns(await future)
[07:30:20]       File "/teamcity/work/56a47d90f832aeb3/venv/lib/python3.9/site-packages/pulumi/output.py", line 216, in run
[07:30:20]         return await cast(Awaitable[U], transformed)
[07:30:20]       File "/teamcity/work/56a47d90f832aeb3/venv/lib/python3.9/site-packages/pulumi_docker/docker.py", line 243, in build_and_push_image
[07:30:20]         build_result = await build_image(base_image_name, path_or_build, log_resource, cache_from)
[07:30:20]       File "/teamcity/work/56a47d90f832aeb3/venv/lib/python3.9/site-packages/pulumi_docker/docker.py", line 444, in build_image
[07:30:20]         raise ResourceError(
[07:30:20]     pulumi_docker.docker.ResourceError: No digest available for image 123456789.dkr.ecr.us-east-2.amazonaws.com/xxx-xxxxxxxxxx-xxxxxxxxx

```

**Comment by ehhthing** [comment]
I can reproduce this on v3.4.1 Python on Github Actions, but not on my M1 Mac.

Downgrading to 3.1.0 fixed it on Github Actions.

**Comment by briandestura** [comment]
> I can reproduce this on v3.4.1 Python on Github Actions, but not on my M1 Mac.
> 
> Downgrading to 3.1.0 fixed it on Github Actions.

Also happens on GCP cloud build, but not when running locally (using M1 Mac)
```
pulumi==3.40.2
pulumi-docker==3.4.1
pulumi-gcp==6.38.0
```
Downgrading to 3.1.0 seems to work for GCP cloud build

**Comment by danielbichuetti** [comment]
I could reproduce it.

> Ubuntu 22.04
> Docker 20.10.22
> Pulumi 3.50.2
> Python 3.10
> pulumi-docker 3.6.1

Downgrading pulumi-docker to 3.1.0 solved the issue.

**Comment by maelp** [comment]
Having this issue too

**Comment by chris-stetter** [comment]
This is still an issue. Any plans to fix this? Downgrading to `3.1.0` blocks upgrading other packages like pulumi-awsx which depends  on pulumi-docker<4.0.0 and >=3.6.1.

**Comment by AaronFriel** [comment]
I'm fairly confident this is resolved in the new implementation of the [Docker Image resource in v4](https://github.com/pulumi/pulumi-docker/releases/tag/v4.0.0)! See our blog post for more info: https://www.pulumi.com/blog/build-images-50x-faster-docker-v4/

If this is still an issue, please file a new issue with the error message(s) and if possible, a minimal program for us to reproduce the issue.
